pkg/storage/ingest/partition_offset_client.go

```diff
@@ -238,3 +249,136 @@ func (p *partitionOffsetClient) FetchTopicLastProducedOffsets(ctx context.Contex
 
 	return offsets, nil
 }
+
+// fetchTopicPartitionsEndOffsets returns the end offset of each partition in the topic. This function returns
+// error if fails to get the offset of any partition (no partial result is returned).
+//
+// The list partitions of the topic is not refreshed for each request, but an in-memory cache is used to reduce
+// the number of requests to Kafka. This is based on the assumption that new partitions don't get added frequently,
+// and a stale view on the existing partitions (but not their offsets) is accepted for a short period of time.
+func (p *partitionOffsetClient) fetchTopicPartitionsEndOffsets(ctx context.Context) (kadm.ListedOffsets, error) {
+	// Fetch the partition IDs. We don't expect partition IDs to change frequently, because we expect
+	// partitions to be pre-provisioned when running Mimir, so use the lookup cache.
+	partitionIDs, err := p.listTopicPartitionIDsWithCache(ctx, 15*time.Second)
+	if err != nil {
+		return nil, err
+	}
+
+	list := kadm.ListedOffsets{
+		p.topic: make(map[int32]kadm.ListedOffset, len(partitionIDs)),
+	}
+
+	// Prepare the request to list offsets.
+	topicReq := kmsg.NewListOffsetsRequestTopic()
+	topicReq.Topic = p.topic
+	for _, partitionID := range partitionIDs {
+		partitionReq := kmsg.NewListOffsetsRequestTopicPartition()
+		partitionReq.Partition = partitionID
+		partitionReq.Timestamp = kafkaOffsetEnd
+
+		topicReq.Partitions = append(topicReq.Partitions, partitionReq)
+	}
+
+	req := kmsg.NewPtrListOffsetsRequest()
+	req.IsolationLevel = 0 // 0 means READ_UNCOMMITTED.
+	req.Topics = []kmsg.ListOffsetsRequestTopic{topicReq}
+
+	// Execute the request.
+	shards := p.client.RequestSharded(ctx, req)
+
+	for _, shard := range shards {
+		if shard.Err != nil {
+			return nil, shard.Err
+		}
+
+		res := shard.Resp.(*kmsg.ListOffsetsResponse)
+		if len(res.Topics) != 1 {
+			return nil, fmt.Errorf("unexpected number of topics in the response (expected: %d, got: %d)", 1, len(res.Topics))
+		}
+		if res.Topics[0].Topic != p.topic {
+			return nil, fmt.Errorf("unexpected topic in the response (expected: %s, got: %s)", p.topic, res.Topics[0].Topic)
+		}
+
+		for _, pt := range res.Topics[0].Partitions {
+			if err := kerr.ErrorForCode(pt.ErrorCode); err != nil {
+				return nil, err
+			}
+
+			list[p.topic][pt.Partition] = kadm.ListedOffset{
+				Topic:       p.topic,
+				Partition:   pt.Partition,
+				Timestamp:   pt.Timestamp,
+				Offset:      pt.Offset,
+				LeaderEpoch: pt.LeaderEpoch,
+			}
+		}
+	}
+
+	return list, nil
+}
+
+// listTopicPartitionIDs returns a list of all partition IDs in the topic.
+func (p *partitionOffsetClient) listTopicPartitionIDs(ctx context.Context) ([]int32, error) {
+	topics, err := p.admin.ListTopics(ctx, p.topic)
+	if err != nil {
+		return nil, err
+	}
+	if err := topics.Error(); err != nil {
```

Comment:
[nit] this threw me off for a minute.

```suggestion
        // Calling .Error() returns only the first error. However, we're requesting only one topic, so there should only be 1 error anyway. 
	if err := topics.Error(); err != nil {
```

Reply:
Not sure adding this comment is much helpful. It returns only 1 error, but we care about the 1st error anyway. This is how all `Error()` functions work in franz-go (they return the 1st error) and this pattern is used everywhere in `pkg/storage/ingest/`.